File: 04_Haskell/workers/src/Behaviour.hs
{-# LANGUAGE ImportQualifiedPost #-}

module Behaviour
  ( Behaviour (..),
    Behaviours,
    none,
    toBit,
  )
where

import Data.Set (Set)
import Data.Set qualified as Set

data Behaviour
  = None_
  | HighPriority
  | LongRunning
  | ResourceIntensive
  | RequiresAffinity
  | Retryable
  | ExactlyOnce
  deriving (Show, Eq, Ord, Enum, Bounded)

type Behaviours = Set Behaviour

none :: Behaviours
none = Set.empty

toBit :: Behaviour -> Int
toBit None_ = 0
toBit HighPriority = 1
toBit LongRunning = 2
toBit ResourceIntensive = 4
toBit RequiresAffinity = 8
toBit Retryable = 16
toBit ExactlyOnce = 32

File: 01_Rust/src/incoming/behaviour.rs
use bitmask_enum::bitmask;

#[bitmask(u8)]
pub enum Behaviour {
    NONE = 0,
    HIGH_PRIORITY = 1 << 0,
    LONG_RUNNING = 1 << 1,
    RESOURCE_INTENSIVE = 1 << 2,
    REQUIRES_AFFINITY = 1 << 3,
    RETRYABLE = 1 << 4,
    EXACTLY_ONCE = 1 << 5,
}

impl Behaviour {
    const BEHAVIOURS: [Behaviour; 6] = [
        Behaviour::HIGH_PRIORITY,
        Behaviour::LONG_RUNNING,
        Behaviour::RESOURCE_INTENSIVE,
        Behaviour::REQUIRES_AFFINITY,
        Behaviour::RETRYABLE,
        Behaviour::EXACTLY_ONCE,
    ];

    pub fn get_all() -> &'static [Behaviour] {
        &Self::BEHAVIOURS
    }
}
